### PR TITLE
WIP: Resource[IO, ExecutionContext] for IOApp

### LINKED ISSUE #337: Reduce ceremony in declaring ExecutionContext for IOApp
The default `IOApp` uses the global execution context for `contextShift` and `timer`.  If we create our own, and want to use it for the `ContextShift` and `Timer` instances, there are three ceremonies to perform:

```scala
  protected implicit val executionContext: ExecutionContextExecutorService =
    ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(8))

  protected override implicit lazy val contextShift: ContextShift[IO] =
    IO.contextShift(executionContext) // ceremony 1

  protected override implicit lazy val timer: Timer[IO] =
    IO.timer(executionContext) // ceremony 2

  def run(args: List[String]): IO[ExitCode] =
    businessLogic
      .guarantee(IO(executionContext.shutdown())) // ceremony 3
      .as(ExitCode.Success)
```

If the trait defined an `ExecutionContext`, we might eliminate ceremonies 1 and 2, but we'd have to be thoughtful about laziness.

A template method, or abstract overrides, might eliminate ceremony 3, but would be unidiomatic for this project.

Going to sleep on this and hope to dream of something cleaner.

### PR DESCRIPTION
**Pull request by rossabaker**
`IOApp` gets an `executionResource: Resource[SyncIO, ExecutionContext]`.  This resource is `use`d to pass an implicit `ExecutionContext` to run.  This change lets us use a non-daemonic `ExeuctionContext` and keep it open for the duration of `run` in a managed and referentially transparent manner.  The choice of `SyncIO` is to guard against a potential future design where the interpretation of `IO` itself may require the resource, so the `Resoure` does not have a circular dependency on the `ConcurrentEffect` instance.

The implicit `ContextShift[IO]` and `Timer[IO]` on `IOApp` are now derived from the implicit `ExeuctionContext`, rather than overridden if an alternate ExecutionContext is used.

By default, we continue to use `ExecutionContext.global`.

Fixes #337

### CODE REVIEW

File: core/jvm/src/main/scala/cats/effect/internals/IOTimerRef.scala [hunk 1]
```diff
@@ -34,7 +34,7 @@ private[effect] trait IOTimerRef {
    * @param ec is the execution context used for actual execution
    *        tasks (e.g. bind continuations)
    */
-  def timer(ec: ExecutionContext): Timer[IO] =
+  implicit def timer(implicit ec: ExecutionContext): Timer[IO] =
```
Comment (rossabaker) [inline core/jvm/src/main/scala/cats/effect/internals/IOTimerRef.scala]: This, and the `ContextShift` change, are controversial.  We pass these around as implicits, but are quite explicit about their construction everywhere but `IOApp`.  The end result in my mind is that we tend to implicitly end up with one in `ExecutionContext.global` than in whatever `ExecutionContext` was created for the app.  I'd like to be talked out of this change and still have the basic mechanism work.

File: core/shared/src/main/scala/cats/effect/IOApp.scala [hunk 2]
```diff
@@ -57,42 +58,14 @@ trait IOApp {
    *
    * @return the [[cats.effect.ExitCode]] the JVM exits with
    */
-  def run(args: List[String]): IO[ExitCode]
+  def run(args: List[String])(implicit executionContext: ExecutionContext): IO[ExitCode]
+
+  protected def executionResource: Resource[IO, ExecutionContext]
```
Comment (rossabaker) [inline core/shared/src/main/scala/cats/effect/IOApp.scala]: Will need scaladoc and mention in tut if we proceed.

**Comment by alexandru** [comment]
@rossabaker going to reiterate my argument at: https://github.com/typelevel/cats-effect/issues/337#issuecomment-417884976

`ExecutionContext` can be boxed in a `Resource` just because `unsafeRunSync()` does not require it for execution. This is a great design for `IO`, but it's not necessarily one that can survive.

1. `Resource[ExecutionContext]` does not work with Monix's `Task`
2. `Resource[ExecutionContext]` would not work for `IO` with a generic `App[F]` that required a `ConcurrentEffect[F]`, which we actually need, because getting a `ConcurrentEffect[IO]` requires a `ContextShift[IO]` as a requirement

In other words boxing `ExecutionContext` in `Resource` is not future proof and depends on a current technicality of `IO` that may or may not survive a major version.

**Comment by alexandru** [comment]
`Resource[ExecutionContext]` cannot work in generic case because `ExecutionContext` needs to be a part of the environment *before* the task gets executed.

If we want the ability to shutdown a thread pool with non-daemonic threads, then this isn't the way to go IMO.

**Comment by rossabaker** [comment]
Not yet proven on a generic effect app, but updated as a proof of concept to cleanly separate the management of the execution context from the `IO` interpreter.

Also backing off making implicit timers and context shifts on the `IO` companion, in favor of in `IOApp`.  Trying to do so is a more drastic change, and caused a curious overload error in ScalaJS.

**Comment by rossabaker** [comment]
`IOAppTests` hangs in coreJS right now.  If this approach seems more acceptable, I'll fix the tests and properly document it ASAP.

**Comment by alexandru** [comment]
@rossabaker I think that the more you try, the uglier it gets 🙂

`Resource[SyncIO, ExecutionContext]` is about usage of a resource via `bracket` in the `SyncIO` context:

- it only works on top of the JVM because we can block via `unsafeRunSync`; this being forced by the call to `use` being synchronous on evaluation
- on top of JS, as you've seen, you have to write an interpreter that basically unboxes that `ExecutionContext`, because `Resource[SyncIO, ?]` is incompatible with what you want to do; it's fine for a utility library, like Linebacker, but in Cats-Effect it kind of smells

The default will also stay `scala.concurrent.ExecutionContext.global`, because that's what most people use as a default, often hard-coded in code and picking another default is risky and has to have a really good reason. The idea of having a default thread-pool that doesn't have daemonic threads might be more popular in other environments, but not in Scala.

Also in Scala our `main` is blocked by the executing `IO` so this isn't your average application that needs to be blocked in order for async actions to run, because our `IO` does that already.

And I just don't see that `ExecutionContext` being a `Resource`.

**Comment by alexandru** [comment]
Btw, there's a third possibility for moving forward ... we can always remove `IOApp`. A micro-library can always fulfill that need while abusing `Resource`. Better safe than sorry.

**Comment by mpilquist** [comment]
Big :-1: on removing IOApp. Doing so will result in lots of copy/pasted implementations at best.

**Comment by SystemFw** [comment]
👎 On removing `IOApp`, at least given how things stand right now.

Basically to make everything work you need `Timer`, `Ec` `ContextShift`, and the mechanics of it are a bit clunky for a variety of reasons. Having `IOApp` makes things work seamlessly for users.

**Comment by rossabaker** [comment]
We already ripped out `fs2.StreamApp` in favor of `IOApp`, and `IOApp` is quite good if you use the global context, and tolerable if you use an alternate daemonic context.  I'd much rather the status quo than lose it entirely.

I could imagine a divergence of `IOApp` between ScalaJS and the JVM.  They're different in Monix.  Has that hurt?

I'm not adamant about using `Resource`, but I still feel like if we can't wrap a managed environment around our `run` method, something is wrong.  I'll think some more on my jog.

**Comment by alexandru** [comment]
Guys, you're more invested in `IOApp` then I am. So whatever you want, I can agree with, just take a day to think about it.

**Comment by rossabaker** [comment]
Closing this in favor of a new iteration.
